consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/FutureHeightTest.java

```diff
@@ -0,0 +1,221 @@
+    peers.forNonProposing(
+        peer ->
+            peer.verifyEstimatedChainHeightEquals(futureHeightBlock.getHeader().getNumber() - 1));
+
+    peers.commitForNonProposing(futureHeightRoundId, futureHeightBlock.getHash());
+
+    peers.verifyNoMessagesReceived();
+    assertThat(context.getCurrentChainHeight()).isEqualTo(0);
+
+    // Add block to chain, and notify system of its arrival.
+    context.getBlockchain().appendBlock(signedCurrentHeightBlock, emptyList());
+    assertThat(context.getCurrentChainHeight()).isEqualTo(1);
+    context
+        .getController()
+        .handleNewBlockEvent(new NewChainHead(signedCurrentHeightBlock.getHeader()));
+
+    final Prepare expectedPrepareMessage =
+        localNodeMessageFactory.createPrepare(futureHeightRoundId, futureHeightBlock.getHash());
+
+    final Commit expectedCommitMessage =
+        new Commit(
+            createSignedCommitPayload(
+                futureHeightRoundId, futureHeightBlock, context.getLocalNodeParams().getNodeKey()));
+
+    peers.verifyMessagesReceived(expectedPrepareMessage, expectedCommitMessage);
+    assertThat(context.getCurrentChainHeight()).isEqualTo(2);
+  }
+
+  @Test
+  public void messagesFromPreviousHeightAreDiscarded() {
+    final Block currentHeightBlock = context.createBlockForProposalFromChainHead(0, 30);
+    final Block signedCurrentHeightBlock =
+        BftHelpers.createSealedBlock(currentHeightBlock, peers.sign(currentHeightBlock.getHash()));
+
+    peers.getProposer().injectProposal(roundId, currentHeightBlock);
+    peers.getNonProposing(0).injectPrepare(roundId, currentHeightBlock.getHash());
+
+    final Prepare expectedPrepareMessage =
+        localNodeMessageFactory.createPrepare(roundId, currentHeightBlock.getHash());
+
+    peers.verifyMessagesReceived(expectedPrepareMessage);
+
+    // Add block to chain, and notify system of its arrival.
+    context.getBlockchain().appendBlock(signedCurrentHeightBlock, emptyList());
+    assertThat(context.getCurrentChainHeight()).isEqualTo(1);
+    context
+        .getController()
+        .handleNewBlockEvent(new NewChainHead(signedCurrentHeightBlock.getHeader()));
+
+    // Inject prepares and commits from all peers for the 'previous' round (i.e. the height
+    // from before the block arrived).
+    peers.prepareForNonProposing(roundId, currentHeightBlock.getHash());
+    peers.commitForNonProposing(roundId, currentHeightBlock.getHash());
+
+    peers.verifyNoMessagesReceived();
+    assertThat(context.getCurrentChainHeight()).isEqualTo(1);
+  }
+
+  @Test
+  public void multipleNewChainHeadEventsDoesNotRestartCurrentHeightManager() {
+    final Block currentHeightBlock = context.createBlockForProposalFromChainHead(0, 30);
+
+    peers.getProposer().injectProposal(roundId, currentHeightBlock);
+    peers.getNonProposing(0).injectPrepare(roundId, currentHeightBlock.getHash());
+
+    peers.clearReceivedMessages();
+
+    // inject a NewHeight FOR THE CURRENT HEIGHT
```

Comment:
comment is meaningless, code says this already

Reply:
done